iree/compiler/Dialect/HAL/Conversion/FlowToHAL/ConvertStreamOps.cpp

```diff
@@ -37,99 +40,477 @@ namespace mlir {
+  ValueAliasingMap valueAliases;
+
+  std::function<void(Value streamValue, Value aliasedValue)> propagateAlias;
+  propagateAlias = [&](Value streamValue, Value aliasedValue) {
+    auto &baseSet = valueAliases[streamValue];
+    baseSet.insert(aliasedValue);
+    auto &aliasedSet = valueAliases[aliasedValue];
+    baseSet.insert(aliasedSet.begin(), aliasedSet.end());
+    aliasedSet.insert(streamValue);
+  };
+
+  // Start with outputs so that we handle tied values that may lead all the way
+  // back up the chain to the stream inputs.
+  auto tiedStreamOp = cast<IREE::TiedOpInterface>(streamOp.getOperation());
+  auto returnOp = cast<IREE::Flow::ReturnOp>(streamBlock->back());
+  for (auto result : llvm::enumerate(streamOp.getResults())) {
+    auto streamValue = returnOp.getOperand(result.index());
+
+    // Tied stream results reuse their stream operand buffer.
+    auto tiedOperandIndex =
+        tiedStreamOp.getTiedResultOperandIndex(result.index());
+    if (tiedOperandIndex.hasValue()) {
+      auto operand = streamBlock->getArgument(tiedOperandIndex.getValue());
+      propagateAlias(streamValue, operand);
+    }
+  }
+
+  for (auto &op : *streamBlock) {
+    auto tiedOp = dyn_cast<IREE::TiedOpInterface>(op);
+    for (auto it : llvm::enumerate(op.getResults())) {
+      auto result = it.value();
+      if (!result.getType().isa<ShapedType>()) continue;
+
+      // Tied results reuse their operand buffer.
+      if (tiedOp) {
+        auto tiedOperandIndex = tiedOp.getTiedResultOperandIndex(it.index());
+        if (tiedOperandIndex.hasValue()) {
+          auto operand = op.getOperand(tiedOperandIndex.getValue());
+          propagateAlias(result, operand);
+        }
+      }
+    }
+  }
+
+  // Inverse the value aliaser->aliasee map so that we have for any particular
+  // value the list of all other values that alias it.
+  for (auto it : valueAliases) {
+    for (auto aliasee : it.second) {
+      for (auto aliaser : it.second) {
+        if (aliaser != aliasee) {
+          valueAliases[aliasee].insert(aliaser);
+        }
+      }
+    }
+  }
+
+  return valueAliases;
+}
+
+//===----------------------------------------------------------------------===//
+// Liveness interval analysis
+//===----------------------------------------------------------------------===//
+
+static constexpr int LIVE_IN = INT_MIN;
+static constexpr int LIVE_OUT = INT_MAX;
+struct LivenessInterval {
+  int start = 0;
+  int end = 0;
+  int ordinal = -1;  // unique per value
+  Value value;
+  bool operator<(const LivenessInterval &rhs) const {
+    return ordinal < rhs.ordinal;
+  }
+};
+using LivenessIntervalMap = DenseMap<Value, LivenessInterval>;
+using LivenessIntervalList = SmallVector<LivenessInterval>;
+
+// Computes the liveness intervals for each value in the stream.
+// Returns a closed range over an arbitrary operation ordering. The LIVE_IN and
+// LIVE_OUT sentinels will be used to indicate values that are live-in and
+// live-out.
```

Comment:
Can you elaborate a bit on what "live-in" and "live-out" mean in this context?

Reply:
added note (concretely it's whether getLiveIn/getLiveOut on Liveness includes the value, but it's representing whether values are captured input arguments or escaping output results)